File: Team_PageFault/minidb/storage/buffer_pool.py
"""Fixed-size page cache. Pages in use are pinned so they can't be evicted
mid-operation; dirty frames are written back before eviction and on shutdown.
Replacement is LRU over the *unpinned* frames, tracked with an OrderedDict whose
most-recently-used entry sits at the end.
"""

import logging
from collections import OrderedDict

from ..constants import BUFFER_POOL_FRAMES
from .disk_manager import DiskManager
from .page import SlottedPage

log = logging.getLogger("minidb.bufferpool")


class BufferPoolFull(RuntimeError):
    """Every frame is pinned, so there's nothing we can evict to make room."""


class Frame:
    __slots__ = ("page_id", "data", "pin_count", "is_dirty")

    def __init__(self, page_id: int, data: bytearray):
        self.page_id = page_id
        self.data = data
        self.pin_count = 0
        self.is_dirty = False


class BufferPool:
    def __init__(self, disk: DiskManager, capacity: int = BUFFER_POOL_FRAMES):
        self.disk = disk
        self.capacity = capacity
        self._frames: "OrderedDict[int, Frame]" = OrderedDict()
        self.hits = 0
        self.misses = 0

    def fetch_page(self, page_id: int) -> Frame:
        if page_id in self._frames:
            self.hits += 1
            self._frames.move_to_end(page_id)        # mark as recently used
            frame = self._frames[page_id]
            frame.pin_count += 1
            return frame
        self.misses += 1
        self._evict_if_needed()
        data = self.disk.read_page(page_id)
        frame = Frame(page_id, data)
        frame.pin_count = 1
        self._frames[page_id] = frame
        return frame

    def new_page(self) -> Frame:
        """Allocate a fresh page on disk, format it, and return it pinned."""
        page_id = self.disk.allocate_page()
        self._evict_if_needed()
        data = bytearray(self.disk.read_page(page_id))
        SlottedPage.init_empty(data)
        frame = Frame(page_id, data)
        frame.pin_count = 1
        frame.is_dirty = True
        self._frames[page_id] = frame
        return frame

    def unpin_page(self, page_id: int, is_dirty: bool) -> None:
        frame = self._frames.get(page_id)
        if frame is None:
            return
        if is_dirty:
            frame.is_dirty = True
        if frame.pin_count > 0:
            frame.pin_count -= 1

    def flush_page(self, page_id: int) -> None:
        frame = self._frames.get(page_id)
        if frame and frame.is_dirty:
            self.disk.write_page(page_id, bytes(frame.data))
            frame.is_dirty = False

    def flush_all(self) -> None:
        for pid in list(self._frames.keys()):
            self.flush_page(pid)

    def _evict_if_needed(self) -> None:
        if len(self._frames) < self.capacity:
            return
        # OrderedDict is in LRU order, so the first unpinned frame is the victim.
        for pid, frame in self._frames.items():
            if frame.pin_count == 0:
                if frame.is_dirty:
                    self.disk.write_page(pid, bytes(frame.data))
                log.debug("evict page %d (dirty=%s)", pid, frame.is_dirty)
                del self._frames[pid]
                return
        raise BufferPoolFull(
            f"all {self.capacity} buffer frames are pinned; cannot evict")

    def stats(self) -> dict:
        total = self.hits + self.misses
        return {
            "hits": self.hits,
            "misses": self.misses,
            "hit_ratio": (self.hits / total) if total else 0.0,
            "resident": len(self._frames),
            "capacity": self.capacity,
        }

File: MiniDB_Projects/Team_EnvPushers/src/storage/buffer_pool.cpp
#include "storage/buffer_pool.hpp"

#include <stdexcept>

namespace minidb {

void BufferPool::touch(PageId pid) {
    auto& f = frames_[pid];
    lru_.erase(f->lru_it);
    lru_.push_back(pid);
    f->lru_it = std::prev(lru_.end());
}

Page* BufferPool::fetch_page(PageId pid) {
    std::lock_guard<std::recursive_mutex> lk(mtx_);
    auto it = frames_.find(pid);
    if (it != frames_.end()) {
        stats_.hits++;
        it->second->pin_count++;
        touch(pid);
        return &it->second->page;
    }
    stats_.misses++;
    evict_if_needed();

    auto frame = std::make_unique<Frame>();
    frame->page = Page(pid, /*init=*/false);
    disk_->read_page(pid, frame->page.raw());
    frame->pin_count = 1;
    lru_.push_back(pid);
    frame->lru_it = std::prev(lru_.end());
    Page* ptr = &frame->page;
    frames_[pid] = std::move(frame);
    return ptr;
}

Page* BufferPool::new_page() {
    std::lock_guard<std::recursive_mutex> lk(mtx_);
    PageId pid = disk_->allocate_page();
    evict_if_needed();

    auto frame = std::make_unique<Frame>();
    frame->page = Page(pid, /*init=*/true);
    frame->pin_count = 1;
    frame->dirty = true;
    lru_.push_back(pid);
    frame->lru_it = std::prev(lru_.end());
    Page* ptr = &frame->page;
    frames_[pid] = std::move(frame);
    return ptr;
}

void BufferPool::unpin_page(PageId pid, bool is_dirty) {
    std::lock_guard<std::recursive_mutex> lk(mtx_);
    auto it = frames_.find(pid);
    if (it == frames_.end()) return;
    if (is_dirty) it->second->dirty = true;
    if (it->second->pin_count > 0) it->second->pin_count--;
}

void BufferPool::flush_frame(Frame& f) {
    if (!f.dirty) return;
    if (before_flush_) before_flush_(f.page.page_id());  // WAL: log first
    disk_->write_page(f.page.page_id(), f.page.raw());
    f.dirty = false;
}

void BufferPool::flush_page(PageId pid) {
    std::lock_guard<std::recursive_mutex> lk(mtx_);
    auto it = frames_.find(pid);
    if (it != frames_.end()) flush_frame(*it->second);
}

void BufferPool::flush_all() {
    std::lock_guard<std::recursive_mutex> lk(mtx_);
    for (auto& [pid, frame] : frames_) flush_frame(*frame);
}

void BufferPool::evict_if_needed() {
    if (frames_.size() < pool_size_) return;
    // Walk LRU order (front = least recent) for the first unpinned frame.
    for (auto it = lru_.begin(); it != lru_.end(); ++it) {
        PageId pid = *it;
        auto fit = frames_.find(pid);
        if (fit->second->pin_count == 0) {
            flush_frame(*fit->second);
            lru_.erase(it);
            frames_.erase(fit);
            stats_.evictions++;
            return;
        }
    }
    throw std::runtime_error("BufferPool full: every frame is pinned");
}

}  // namespace minidb
